### PR TITLE
Adding map ex

### PR DESCRIPTION
**Pull request by Frenchris**

### CODE REVIEW

File: exercise/map.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+*/
+let arr1 = [10, -10, 20, -95, 86]
+///*/ // ⚡
+
+///*/// ⚡
+
+let arrRandom1 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+
+for (let i = 0; i < 10; i++) {
+  let value = Math.round(Math.random() * (10 - -10 - 10))
```
Comment (kigiri) [inline exercise/map.js]: 🥇 `(10 - -10 - 10)` most complicated way of writting `10` I have seen yet.

File: exercise/map.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  arrRandom1.push(value)
```
Comment (kigiri) [inline exercise/map.js]: If you need random values in your tests, it's best not to expose them outside of tests.
For that you can use test context:

```js
t((_, ctx) => ctx.arrRandom1 = Array(10).fill().map(() => Math.round(Math.random() * 10)))
```
you can then use your context like so:
```js
t(({ eq }, ctx) => eq(fn(ctx.arrRandom1)[0], check(arr[0]))
```

File: exercise/map.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+The use of `for` and `while` is forbidden for this exercise
+
+### Notions
+
+- Map
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
+
+*/
+let arr1 = [10, -10, 20, -95, 86]
```
Comment (kigiri) [inline exercise/map.js]: this array is so small they can just write the solution doing the math in theyr head.
You don't even need JS for that haha

File: exercise/map.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+### Instructions
+
+Create an array named 'arr2' that contains the values of 'arr1'
+with the following modifications applied to each of the values:
+1) They have to be multiplied by 2
+2) Then the results have to be subtracted by 4
+3) finally, the negative values have to be replaced by null
+then do the same with arrRandom1 by returning arrRandom2
+The use of `for` and `while` is forbidden for this exercise
```
Comment (kigiri) [inline exercise/map.js]: I think this should be a functions, and even more that we should have a few different use cases.
Not just 1 use of map.

I also don't thing this is a very demonstrative usecase of map.

File: exercise/map.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  for (let i of arrRandom1) {
+    if (arrRandom1[i] >= 2)
+      if (arrRandom2[i] !== check(arrRandom1[i])) return false
+    if (arrRandom1[i] < 2) if (arrRandom2[i] !== null) return false
+    return true
+  }
+})
+
+t(({ code }) => !code.includes('for'))
+t(({ code }) => !code.includes('while'))
```
Comment (kigiri) [inline exercise/map.js]: I think a better way to ensure they used map would be to "monkey patch" map:

```js
t((_, ctx) => {
  ctx.mapCall = []
  const { map } = Array.prototype
  Array.prototype.map = function () {
    ctx.mapCalls.push([this, ...arguments])
    return map.apply(this, arguments)
  }
})
```
now this wierd shitty code allow to wrap the *classic* `[].map` in a function that save the arguments of each calls.

so in your tests you can check in the context if they used map or not, and try to mimic the call:
```js
t(({ eq }, ctx) => eq(fn([10, -2, 4]), [16, null, 0]) // 1st test execute it
t(({ eq }, ctx) => eq(ctx.mapCalls[ctx.mapCalls.length - 1][0], [10, -2, 4]) // check that map was called on this array
```

**Review by kigiri — CHANGES_REQUESTED** [review]

File: exercise/map.js (hunk 2 above)
Comment (Frenchris) [reply exercise/map.js]: Noted. Declaring them in the setup now.

File: exercise/map.js (hunk 3 above)
Comment (Frenchris) [reply exercise/map.js]: I wanted to show an easy "example" one along with a random one.

File: exercise/map.js (hunk 4 above)
Comment (Frenchris) [reply exercise/map.js]: reworked the subject to ask for functions.

File: exercise/map.js (hunk 5 above)
Comment (Frenchris) [reply exercise/map.js]: In use now. I transmitted the knowledge to Afonso.

**Review by kigiri — COMMENTED** [review]
not fund of `map1`, `map2` and `map3` names

**Review by kigiri — CHANGES_REQUESTED** [review]
while having `choumi` arround is fun, I dislike including us in the subject, I'd like them to be more neutral.

If we think about when to use map:
- Change the shape of an array (ex: *(from array of user `object` to array of name `string`)*)
	```js
	[{ name: 'vanessa' }, { name: 'gaetan' }].map(user => user.name)
	```

- Apply a computations on an array of numbers
	this is where you did `map1` but really what's the point of your calculation, it's wierd and confusing.
	```js
	[0, 10.10, 87, 55, 80].map(fahrenheitToCelsius)
	```
	something like that achieve the same goal and is actually relatable.
    I mean I could see how this is usefull.

- Changing only some elements in an array:
	```js
	const changePassword = (users, id, password) =>
		users.map(user => user.id === id ? {...user, password } : user)
	```

The problem I have with your exemples is that they feel far from data we would have in a program or realife usecase.

please make sure to add tests with `Object.freeze` so we confirm that they didn't edit the array :)

**Comment by Frenchris** [comment]
Did an update of the subject based on the last two comments. Awaiting feedback.

File: exercise/using-map.js [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+  {
+    city: 'Los Angeles',
+    temperature: '  101 °F   ',
+  },
+  {
+    city: 'San Francisco',
+    temperature: ' 84 ° F   ',
+  },
+]
+```js
```
Comment (kigiri) [inline exercise/using-map.js]: it's:
\`\`\`js
\`\`\`
not:
\`\`\`
\`\`\`js

File: exercise/using-map.js [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+```
+['Los Angeles', 'San Francisco']
+```js
+
+-Create a function named 'upperCasingStates' which takes an array of strings
+and which Upper Case each words of a string.
+The function returns then an array of strings.
+
+Example:
+['alabama', 'new jersey']
```
Comment (kigiri) [inline exercise/using-map.js:36]: no js here ?

File: exercise/using-map.js [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+[
+  {
+    city: 'Los Angeles',
+    temperature: '101°F',
+  },
+  {
+    city: 'San Francisco',
+    temperature: '84°F',
+  },
+]
```
Comment (kigiri) [inline exercise/using-map.js:78]: inline those:
```js
[
  { city: 'Los Angeles', temperature: '101°F' },
  { city: 'San Francisco', temperature: '84°F' },
]
```
that's plenty readable.

File: exercise/using-map.js [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+['38°Celsius in Pasadena, California']
+```js
+
+### Notions
+
+- Map
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
+
+*/
```
Comment (kigiri) [inline exercise/using-map.js]: `*/` is extraneous

File: exercise/using-map.js [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+```
+['38°Celsius in Pasadena, California']
+```js
+
+### Notions
+
+- Map
+
+https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
```
Comment (kigiri) [inline exercise/using-map.js]: only links, and from devdocs if possible.

File: module/call-me-maybe.js [hunk 11]
```diff
@@ -1,9 +1,20 @@
...
+  // foldl: { feedback: true },
+  // foldr: { feedback: true },
+  // reduce: { feedback: true },
+  // 'do-your-own-reduce': { feedback: true },
+  // every: { feedback: true },
+  // 'do-your-own-every': { feedback: true },
+  // curry: { feedback: true },
+  // curry2: { feedback: true },
+  // 'do-your-own-curry': { feedback: true },
+  // some: { feedback: true },
```
Comment (kigiri) [inline module/call-me-maybe.js]: rebase with gh-pages and remove the comments durring the rebase conflicts.

**Review by kigiri — CHANGES_REQUESTED** [review]

**Review by kigiri — APPROVED** [review]

**Pull request merged**